### DISCUSSION TITLE
The Diagnostic Report

### DISCUSSION THREAD

**Discussion by D8H** [4ian/GDevelop#2919] [Enhancement ideas (and their technical discussions)]
This is a mock-up of the Diagnostic Report to open the discussion.

* The user can jump to a given issue using the loupe.
  * Here, it would open the object editor to show the unfilled property.
* The user can fix a batch of issues
* When it's not possible to fix an issue automagically, the description can help the user to understand how to fix it.

![image](https://user-images.githubusercontent.com/2611977/129457205-d524bb24-cb38-4978-b33d-d635498689d0.png)

More details about the Diagnostic Report can be found on the ticket:
https://trello.com/c/IiLgNR16/462-add-a-diagnostic-report-to-warn-about-potential-issues-in-the-game-and-show-them-in-the-events-sheet

**Comment by Silver-Streak** [comment]
I love the style here. The only thing I would recommend is maybe color coding the problems? 
That may require extra work, but I'd think something similar to the following could be very helpful to users:

- "This will break loading" as red (Missing resources is a common one)
- "This may break the game" as orange (Unsupported instruction/codec)
- "This may cause unexpected/unwanted behavior" as yellow (Unsupported codec/unused resources/trigger once in a for each)

**Reply by 4ian** [reply]
Yep we could have varying degrees of problems. The red one breaking loading would also prevent a preview (if you try to export/preview, the diagnostic manager would popup with these problems highlighted).

I was also thinking that the diagnostic report would maybe need a "scan project" button to run all the checks (though I'm unsure if this is a good UX or a workaround against the fact that not all reports may be immediately visible). 
For example, the reports about a required parameter not being filled in an action may only be discovered when generating the events - i.e: at preview or export. In this situation, it would be a good idea to have something to run these checks on demand, rather than waiting for a preview/export.

**Comment by tristanbob** [comment]
Here is the syntax problem that I wasted a lot of time today (and has bitten me in the past):

```
Quotations were used in a field that should contain a variable name 
-- Triggered Event:  Toggle the boolean value of boolean variable "MyBool" 
-- Recommended change:  Toggle the boolean value of boolean variable MyBool
```

This would be a good check to add to the diagnostic report.

**Reply by 4ian** [reply]
True. I wonder if we could also add this to the variable fields so that we can directly warn "You entered a quote on this variable name, are you sure about this?"

**Reply by 4ian** [reply]
Partly fixed in #3652 - should help this kind of mistake :)
